hw12-webserver

alexsysoy opened the pull request:
Добавил задание по webserver

alexsysoy committed:
Добавил задание по webserver

stvort commented on hw12-webserver/src/main/java/db/repository/ClientRepository.java:
```diff
@@ -0,0 +1,10 @@
+
+import java.util.List;
+
+public interface ClientRepository {
```
У нас в ДЗ по гибернейту вроде не было такой абстракции

alexsysoy replied:
Не было, это я её добавил, как объект, скрывающий работу с дб. Убрал

stvort commented on hw12-webserver/src/main/java/db/repository/ClientRepositoryImpl.java:
```diff
@@ -0,0 +1,48 @@
+public class ClientRepositoryImpl implements ClientRepository {
+    public static final String HIBERNATE_CFG_FILE = "hibernate.cfg.xml";
+
+    private final DBServiceClient dbServiceClient;
```
При этом репозиторий не может использовать сервисы. Это разные слои приложения. Дао/репозиторий это слой работы с данными. Слои пересекаются как-то так (в направлении стрелок):
Представление --использует--> Сервисный слой  --использует--> Слой хранения --использует--> Слой доменных объектов.

alexsysoy replied:
Убрал репозиторий, теперь работает как и задумывалось ClientService

stvort commented on hw12-webserver/src/main/java/db/repository/ClientRepositoryImpl.java:
```diff
@@ -0,0 +1,48 @@
+    private final DBServiceClient dbServiceClient;
+
+    public ClientRepositoryImpl() {
+        dbServiceClient = createDBServiceClient();
```
Зависимости лучше передавать извне. Иначе это нарушение DIP из SOLI**D**

alexsysoy replied:
Убрал абстракцию ClientRepository, добавил передачу зависимостей через конструкторы

stvort commented on hw12-webserver/src/main/java/db/repository/ClientRepositoryImpl.java:
```diff
@@ -0,0 +1,48 @@
+        var transactionManager = new TransactionManagerHibernate(sessionFactory);
+        var clientTemplate = new DataTemplateHibernate<>(Client.class);
+        return new DbServiceClientImpl(transactionManager, clientTemplate);
+    }
```
А это уже дополнительно нарушает SRP из **S**OLID

alexsysoy replied:
Убрал этот функционал в класс Demo

stvort commented on hw12-webserver/src/main/java/webserver/servlet/ClientsServlet.java:
```diff
@@ -0,0 +1,52 @@
+
+    public ClientsServlet(TemplateProcessor templateProcessor) {
+        this.templateProcessor = templateProcessor;
+        this.clientRepository = new ClientRepositoryImpl();
```
И сюда тоже лучше передавать зависимости извне. У нас в примере не зря все создание всех компонентов приложения было в `main`

alexsysoy replied:
Изменил на передачу зависимостей извне

stvort commented on hw12-webserver/src/main/java/webserver/servlet/ClientsServlet.java:
```diff
@@ -0,0 +1,52 @@
+    @Override
+    protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws IOException {
+        Map<String, Object> paramsMap = new HashMap<>();
+        paramsMap.put("clients", clientRepository.getAllClientFromDb());
```
Кстати в реальном интерфейсе репозитория не могло быть в названиях методов "FromDb", т.к. основная задача данного паттерна, это срывать за собой источник данных, а "FromDb" его выдает с головой

alexsysoy replied:
Заменил названия методов

stvort commented on hw12-webserver/src/main/resources/hibernate.cfg.xml:
```diff
@@ -0,0 +1,23 @@
+    <property name="hibernate.format_sql">false</property>
+    <property name="hibernate.generate_statistics">true</property>
+
+    <property name="hibernate.hbm2ddl.auto">create</property>
```
Лучше, чтобы за создание схемы и наполнение данными отвечали миграции

alexsysoy replied:
Добавил миграции

alexsysoy committed:
Правки по замечаниям:
1. Убрал лишнбб абстракцию ClientRepository
2. Добавил передачу зависимостей через конструктор
3. Убрал несекьюрные названия методов
4. Добавил миграции

stvort commented on hw12-webserver/src/main/java/webserver/Demo.java:
```diff
@@ -0,0 +1,90 @@
+        return new DbServiceClientImpl(transactionManager, clientTemplate);
+    }
+
+    private static void flywayMigrations() {
+        log.info("db migration started...");
+        var flyway = Flyway.configure()
+                .dataSource(URL, USER, PASSWORD)
+                .locations("classpath:/migration")
+                .load();
+        flyway.migrate();
+        log.info("db migration finished.");
+        log.info("***");
+    }
```
За это у нас сервис отвечал в ДЗ по гиберу. В `main` не место такого рода логике. Только сборка и запуск

alexsysoy replied:
Понял, убрал

stvort commented on hw12-webserver/src/main/java/webserver/Demo.java:
```diff
@@ -0,0 +1,90 @@
+        var dbUserName = configuration.getProperty("hibernate.connection.username");
+        var dbPassword = configuration.getProperty("hibernate.connection.password");
+
+        new MigrationsExecutorFlyway(dbUrl, dbUserName, dbPassword).executeMigrations();
```
Вот он кстати. Т.е. мы два раза запускаем миграции

alexsysoy replied:
Запустил эти миграции

stvort commented on hw12-webserver/src/main/java/webserver/servlet/ClientsServlet.java:
```diff
@@ -0,0 +1,51 @@
+        String name = req.getParameter("name");
+        String address = req.getParameter("address");
+
+        Client client = dbServiceClient.saveClient(new Client(null, name, new Address(null, address), new ArrayList<>()));
```
А как же телефон(ы)?

alexsysoy replied:
Добавил телефоны

stvort commented on hw12-webserver/src/main/java/db/crm/dbmigrations/MigrationsExecutorFlyway.java:
```diff
@@ -0,0 +1,24 @@
+    public MigrationsExecutorFlyway(String dbUrl, String dbUserName, String dbPassword) {
+        flyway = Flyway.configure()
+                .dataSource(dbUrl, dbUserName, dbPassword)
+                .locations("classpath:/db/migration")
```
Ага. Видимо эта штука не привела к появлению схемы БД и мы добавили еще один метод. Если так, то тут просто путь к миграциям не верный

alexsysoy replied:
Всё верно, исправил

alexsysoy committed:
Исправления по замечаниям
